### PR TITLE
Manage multiple organisations

### PR DESCRIPTION
**Pull request by MartinBentley**
### Context

Depends on https://github.com/DFE-Digital/manage-courses-api/pull/51

### Changes proposed in this pull request:

The user journey is now driven by organization.
If a user belongs to multiple organisations then the "Organisations" page becomes the landing page.  This gives the user the option to select an organization on which the rest of the user journey is based.
If the user belongs to just one organization then they are taken directly to the "courses" page for that organization.
When the user selects a course to view then the user is taken to the course "details" for that course.
Note: When the user has multiple organisations then the bread-crumb shows "Organisations". When the user has one organization then "Organisations" is dropped from the bread-crumb
### Guidance to review
landingpage

viewc
![coursespage

course

courses page If a user were to edit the url with invalid organisation

coursespageeditedurl

### CODE REVIEW

**Comment by dankmitchell** [comment]
Can you give some context to this please @MartinBentley. Prehaps a screenshot and guidance on how to review....

**Comment by fofr** [comment]
Just to check – if I were to edit the URL to load a different organisation, what happens? I'm guessing it should be a 401/403 response, it definitely should not show another organisation’s courses that I don't have access to.

**Comment by MartinBentley** [comment]
The organization courses screen only shows courses for organisations the user has access to.  If you edit the url and put in the id of an organization that the user doesn't have access to (or if the organization does not exist) then 0 course are returned

**Comment by fofr** [comment]
@MartinBentley Can you show me what a user sees if they enter an organisation they don't have access to?

File: src/ui/Controllers/CourseController.cs [hunk 1]
```diff
@@ -48,7 +47,7 @@ public CourseController(ManageApi manageApi)
                     UcasCode = course.UcasCode,
                     AgeRange = courseDetail.AgeRange,
                     Route = variant.ProgramType,
-                    Qualifications= variant.ProfPostFlag,
+                    Qualifications = variant.ProfPostFlag,
```
Comment (dankmitchell) [inline src/ui/Controllers/CourseController.cs]: Is there some sort of linter we can add to catch these going forward?

File: src/ui/Views/Organisations/Index.cshtml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿@model OrganisationListViewModel
+
+@{
+    Layout = "_Layout";
```
Comment (dankmitchell) [inline src/ui/Views/Organisations/Index.cshtml]: Can we add a page title here, please? `@organisation.OrganisationName`

File: ManageCoursesUi.sln.DotSettings.user [hunk 3]
```diff
@@ -0,0 +1,4 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/Environment/AssemblyExplorer/XmlDocument/@EntryValue">&lt;AssemblyExplorer&gt;&#xD;
+  &lt;Assembly Path="C:\Users\cex\.nuget\packages\smartbreadcrumbs\1.0.1\lib\netcoreapp2.0\SmartBreadcrumbs.dll" /&gt;&#xD;
```
Comment (timabell) [inline ManageCoursesUi.sln.DotSettings.user]: This has your username in it. I think the *.user files should be ignored in the `.gitignore` file

File: src/ui/Controllers/CourseController.cs (hunk 1 above)
Comment (timabell) [reply src/ui/Controllers/CourseController.cs]: There's StyleCop but I don't like it much.

File: src/ui/Controllers/HomeController.cs [hunk 4]
```diff
@@ -17,10 +18,22 @@ public HomeController(ManageApi manageApi)
...
+                return this.RedirectToAction("Index", "Courses", new {organisationId = orgs[0].OrganisationId });
+            }
+
+            if (orgs.Count() > 1)
+            {
+                return this.RedirectToAction("Index", "Organisations");
+            }
 
-            return this.RedirectToAction("Index", "Courses");
+            return null;
```
Comment (timabell) [inline src/ui/Controllers/HomeController.cs]: I assume this is the failure case in which case it should throw an exception

File: src/ui/Views/_ViewImports.cshtml [hunk 5]
```diff
@@ -1,4 +1,4 @@
 @using GovUk.Education.ManageCourses.ApiClient
 @using GovUk.Education.ManageCourses.Ui.ViewModels
 @using GovUk.Education.ManageCourses.Ui.Helpers
-@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
```
Comment (timabell) [inline src/ui/Views/_ViewImports.cshtml]: Any reason for this diff? Looks like diff noise

File: src/ui/Views/Organisations/Index.cshtml (hunk 2 above)
Comment (MartinBentley) [reply src/ui/Views/Organisations/Index.cshtml]: All done

**Comment by timabell** [comment]
Running locally with custom data I got two orgs on the first page, but the links were the same.

File: src/ui/Controllers/CourseController.cs [hunk 6]
```diff
@@ -20,11 +20,10 @@ public CourseController(ManageApi manageApi)
         {
             _manageApi = manageApi;
         }
-
         [Route("{accreditingProviderId=self}/{courseTitle}/{ucasCode}")]
-        public async Task<IActionResult> Variants(string accreditingProviderId, string courseTitle, string ucasCode)
+        public async Task<IActionResult> Variants(string accreditingProviderId, string courseTitle, string ucasCode, string organisationId)
```
Comment (defong) [inline src/ui/Controllers/CourseController.cs]: What is the expect url in the browser?
@fofr 
`{accreditingProviderId=self}/{courseTitle}/{ucasCode}?organisationId=123`
site.com/course/self/coursetitle/ucascode/?organisationId=123

I think it should be
site.com/course/123/self/coursetitle/ucascode/
as we already have
site.com/course/123

File: src/ui/Controllers/CourseController.cs [hunk 7]
```diff
@@ -75,10 +74,12 @@ public CourseController(ManageApi manageApi)
                         };
                     })
                 };
-
+            var orgs = await _manageApi.GetOrganisations();
             var viewModel = new FromUcasViewModel//TODO change view model to show on course variant
```
Comment (defong) [inline src/ui/Controllers/CourseController.cs]: Is that comment still relevant?

File: src/ui/Controllers/CoursesController.cs [hunk 8]
```diff
@@ -19,16 +19,20 @@ public CoursesController(ManageApi manageApi)
...
+            var data = await _manageApi.GetOrganisationCoursesTotal(organisationId);
             var model = new CourseListViewModel
             {
                 Courses = courses,
-                TotalCount = data.TotalCount
+                TotalCount = data.TotalCount,
+                MultipleOrganisations = orgs.Count() > 1
             };
-            return View(model);
+            return View("Index", model);
```
Comment (defong) [inline src/ui/Controllers/CoursesController.cs]: Is there an actual need for string "index"? 
`return View(model);` that should suffice

File: src/ui/Controllers/OrganisationsController.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    [Route("organisations")]
+    public class OrganisationsController : CommonAttributesControllerBase
+    {
+        private readonly ManageApi _manageApi;
+
+        public OrganisationsController(ManageApi manageApi)
+        {
+            _manageApi = manageApi;
+        }
+        public async Task<IActionResult> Index()
```
Comment (defong) [inline src/ui/Controllers/OrganisationsController.cs]: Will this be a post as well?

File: src/ui/Controllers/CourseController.cs (hunk 6 above)
Comment (MartinBentley) [reply src/ui/Controllers/CourseController.cs]: Url fixed as suggested

File: src/ui/Controllers/CourseController.cs (hunk 7 above)
Comment (MartinBentley) [reply src/ui/Controllers/CourseController.cs]: removed

File: src/ui/Views/Courses/Index.cshtml [hunk 10]
```diff
@@ -2,11 +2,24 @@
 
 @{
     Layout = "_Layout";
+    ViewBag.Title = Model.Courses.OrganisationName;
+}
+@if (Model.MultipleOrganisations)
```
Comment (defong) [inline src/ui/Views/Courses/Index.cshtml]: With regards to this you may want to look at this implementation

https://github.com/DFE-Digital/search-and-compare-ui/blob/23738eb962d982007a01d63e6558a1cf1ee2e5bb/src/Views/Results/Index.cshtml#L46

File: src/ui/Controllers/CoursesController.cs [hunk 11]
```diff
@@ -19,14 +19,18 @@ public CoursesController(ManageApi manageApi)
...
-        public async Task<IActionResult> Index()
+        [Route("{ucasCode}")]        
+        public async Task<IActionResult> Index(string ucasCode)
         {
-            var courses = await _manageApi.GetCourses();
-            var data = await _manageApi.GetOrganisationCoursesTotal();
+            var courses = await _manageApi.GetCoursesByOrganisation(ucasCode);
+            var orgs = await _manageApi.GetOrganisations();
+
+            var data = await _manageApi.GetOrganisationCoursesTotal(ucasCode);
```
[the remark this answers is not in the document]
Comment (MartinBentley) [reply src/ui/Controllers/CoursesController.cs]: Yes we dont need this now. It's gone

[the remark this answers is not in the document]
Comment (MartinBentley) [reply src/ui/Controllers/CoursesController.cs]: Oops! We do need it. However it's calculating incorrectly so needs to be looked at

File: src/ui/Views/Courses/Index.cshtml (hunk 10 above)
Comment (defong) [reply src/ui/Views/Courses/Index.cshtml]: That's fine

**Pull request merged**